File: spec/javascripts/jsx/context_cards/StudentContextTraySpec.js
/*
 * Copyright (C) 2016 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import $ from 'jquery'
import React from 'react'
import ReactDOM from 'react-dom'
import TestUtils from 'react-dom/test-utils'
import StudentContextTray from '@canvas/context-cards/react/StudentContextTray'

QUnit.module('StudentContextTray', hooks => {
  let tray
  const courseId = '1'
  const studentId = '1'

  hooks.beforeEach(() => {
    tray = TestUtils.renderIntoDocument(
      <StudentContextTray
        courseId={courseId}
        studentId={studentId}
        returnFocusTo={() => {}}
        data={{loading: true}}
      />
    )
  })
  hooks.afterEach(() => {
    if (tray) {
      const componentNode = ReactDOM.findDOMNode(tray)
      if (componentNode) {
        ReactDOM.unmountComponentAtNode(componentNode.parentNode)
      }
    }
    tray = null
  })

  test('tray should set focus back to the result of the returnFocusTo prop', () => {
    $('#fixtures').append('<button id="someButton"><button>')
    const component = TestUtils.renderIntoDocument(
      <StudentContextTray
        courseId={courseId}
        studentId={studentId}
        data={{loading: true}}
        returnFocusTo={() => [$('#someButton')]}
      />,
      document.getElementById('fixtures')
    )

    const fakeEvent = {
      preventDefault() {}
    }
    component.handleRequestClose(fakeEvent)
    ok(document.activeElement === document.getElementById('someButton'))
  })

  QUnit.module('analytics button', () => {
    const user = {
      short_name: 'wooper',
      enrollments: []
    }

    const course = {
      permissions: {
        view_analytics: true
      },
      submissionsConnection: {edges: []}
    }

    const analytics = {
      participations: {level: 2},
      page_views: {level: 3}
    }

    test('it renders with analytics data', () => {
      const userWithAnalytics = {...user, analytics}

      tray = TestUtils.renderIntoDocument(
        <StudentContextTray
          courseId={courseId}
          studentId={studentId}
          returnFocusTo={() => {}}
          data={{
            loading: false,
            user: userWithAnalytics,
            course
          }}
        />,
        document.getElementById('fixtures')
      )
      const quickLinks = tray.renderQuickLinks(userWithAnalytics, course)
      const children = quickLinks.props.children.filter(quickLink => quickLink !== null)

      // This is ugly, but getting at the rendered output with a portal
      // involved is also ugly.
      ok(children[0].props.children.props.href.match(/analytics/))
    })

    test('it renders analytics 2 button (only) if the tool is installed', () => {
      const userWithAnalytics = {...user, analytics}

      tray = TestUtils.renderIntoDocument(
        <StudentContextTray
          courseId={courseId}
          studentId={studentId}
          returnFocusTo={() => {}}
          data={{
            loading: false,
            user: userWithAnalytics,
            course
          }}
          externalTools={[
            {
              title: 'Analytics Beta',
              base_url:
                'http://example.com/courses/1/external_tools/29?launch_type=student_context_card',
              tool_id: 'fd75124a-140e-470f-944c-114d2d93bb40',
              icon_url: null,
              canvas_icon_class: 'icon-analytics'
            }
          ]}
        />,
        document.getElementById('fixtures')
      )
      const quickLinks = tray.renderQuickLinks(userWithAnalytics, course)
      const children = quickLinks.props.children.filter(quickLink => quickLink !== null)

      equal(children.length, 1)
      equal(
        children[0][0].props.children.props.href,
        'http://example.com/courses/1/external_tools/29?launch_type=student_context_card&student_id=1'
      )
    })

    test('it does not render without analytics data', () => {
      tray = TestUtils.renderIntoDocument(
        <StudentContextTray
          courseId={courseId}
          studentId={studentId}
          returnFocusTo={() => {}}
          data={{
            loading: false,
            user,
            course
          }}
        />,
        document.getElementById('fixtures')
      )
      const quickLinks = tray.renderQuickLinks(user, course)
      const children = quickLinks.props.children.filter(quickLink => quickLink !== null)
      ok(children.length === 0)
    })
  })
})

File: spec/selenium/admin/analytics_2/student_context_tray_spec.rb
# frozen_string_literal: true

# Copyright (C) 2019 - present Instructure, Inc.

# This file is part of Canvas.

# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.

# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.

# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

require_relative '../../common'
require_relative '../pages/course_page.rb'
require_relative '../pages/student_context_tray_page.rb'
require_relative '../../../factories/analytics_2_tool_factory.rb'

describe "analytics in Canvas" do
  include_context "in-process server selenium tests"
  include CourseHomePage
  include StudentContextTray
  include Factories

  context "Analytics 2.0 LTI installed" do
    before :once do
      @admin = account_admin_user(:active_all => true)
      # Analytics1.0 is enabled for all tests by default
      @admin.account.update(allowed_services: "+analytics")
      # add the analytics 2 LTI to the account
      analytics_2_tool_factory
      @tool_id = @admin.account.context_external_tools.first.id
      # create a course, @teacher and student in course
      @course = course_with_teacher(
        :account => @admin.account,
        :course_name => "A New Course",
        name: 'Teacher1',
        :active_all => true
      ).course
      @student = student_in_course(
        :course => @course,
        :name => "First Student",
        :active_all => true
      ).user
    end

    describe "student context tray for teacher role" do
      context "with A2 FF enabled" do
        before :each do
          @course.root_account.enable_feature!(:analytics_2)
          user_session(@teacher)

          visit_course_people_page(@course.id)
          course_user_link(@student.id).click
          wait_for_student_tray
        end

        it "displays Analytics 2 button on Student Tray" do
          skip "Flakey spec. Fix via LA-849"
          expect(student_tray_quick_links.text).to include('Analytics 2')
        end
      end

      context "with A2 FF disabled" do
        before :each do
          @course.root_account.disable_feature!(:analytics_2)
          user_session(@teacher)

          visit_course_people_page(@course.id)
          course_user_link(@student.id).click
          wait_for_student_tray
        end

        it "displays Analytics 1 button on Student Tray" do
          skip "Flakey spec. Fix via LA-849"
          expect(student_tray_quick_links.text).to include('Analytics')
          expect(student_tray_quick_links.text).not_to include('Analytics 2')
        end
      end

      context "with permissions" do
        context "with A2 FF disabled and view_analytics permission disabled" do
          before :each do
            @course.account.role_overrides.create!(:permission => :view_analytics, :role => teacher_role, :enabled => false)
            @course.root_account.disable_feature!(:analytics_2)
            user_session(@teacher)

            visit_course_people_page(@course.id)
            course_user_link(@student.id).click
            wait_for_student_tray
          end

          it "does not display Analytics 1 button" do
            skip "Flakey spec. Fix via LA-849"
            expect(student_tray_quick_links.text).not_to include('Analytics')
          end
        end

        context "with A2 FF enabled and view_all_grades disabled" do
          before :each do
            skip "Flakey spec. Fix via LA-849"

            @course.account.role_overrides.create!(:permission => :view_all_grades, :role => teacher_role, :enabled => false)
            @course.root_account.enable_feature!(:analytics_2)
            user_session(@teacher)

            visit_course_people_page(@course.id)
            course_user_link(@student.id).click
            wait_for_student_tray
          end

          it "does not display Analytics 2 button" do
            expect(student_tray_quick_links.text).not_to include('Analytics 2')
          end
        end
      end
    end
  end
end